### PR TITLE
C++: Fix FPs from cpp/wrong-type-format-argument

### PR DESCRIPTION
**Pull request by geoffw0**
The `cpp/wrong-type-format-argument` query produces false positive results when the snapshot was built on Linux but the code was written specifically for Windows, or vice-versa.  In principle the snapshot should be built on the platform(s) it's intended for, it practice this problem crops up both on LGTM and in the SAMATE tests.  The solution is to ignore results that would be correct on the other platform (taking care to still generate the most accurate expected type we can in the alert message).

SAMATE results are here: https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp-detailed/97/ (there are reductions in FPs for CWE-134 and several others, but unfortunately we also lose some coincidental TPs in CWE-534 and CWE-535; we could have 100% TPs on both of those queries by enabling `cpp/cleartext-storage-file` though, or on one of them by enabling `cpp/system-data-exposure`, so that would be the right way to repair those results).

Example LGTM differences (selected projects for relevance): before https://lgtm.com/query/5589321904419439737/; after https://lgtm.com/query/8537753305478530046/ (we lose 35 FPs on ChakraCore)

I've also removed two of the `pragma[noopt]`s in this query (one dating back to 2018 or before, the other introduced in this branch by copying it) and simplified some bits to improve code maintainability and readability.  On a variety of snapshots I've only seen timings increase or decrease by a few seconds.

### CODE REVIEW

**Comment by geoffw0** [comment]
I've attempted to start a CPP-Differences run here: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2129/

**Comment by geoffw0** [comment]
The CPP-Differences job shows we have lost a lot of results for `cpp/wrong-type-format-argument` on `ssh/github/codeql-samate`.  This is expected and good.

Analysis times show a total slowdown from 93s to 113s for `WrongTypeFormatArguments.ql`.  The biggest slowdowns are in `g/torvalds/linux` and `g/kamailio/kamailio` (5s each).  I can't reproduce the former locally (could be noise???), I seem to be able to reproduce the latter and am (briefly) investigating ... but regardless this is not an especially slow query according to CPP-Differences.

**Comment by geoffw0** [comment]
I've just pushed a fix for the (small) performance regression.

File: cpp/ql/src/semmle/code/cpp/File.qll [hunk 1]
```diff
@@ -358,6 +358,11 @@ class File extends Container, @file {
   string getShortName() { files(underlyingElement(this), _, result, _, _) }
 }
 
+/**
+ * Holds if any file was compiled by a Microsoft compiler.
+ */
+cached predicate anyFileCompiledAsMicrosoft() { any(File f).compiledAsMicrosoft() }
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/File.qll:364]: Adding a cached stage just for this seems heavy-handed. Isn't it enough to write `pragma[noinline]`?

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/File.qll]: That doesn't seem to be enough, but I've found a solution improving `compiledAsMicrosoft()` and removing `cached`.

**Comment by rdmarsh2** [comment]
Is there a better way to tell what standard library the code is actually being compiled against? I know ChakraCore is compiled with the [Platform Abstraction Layer](https://github.com/microsoft/pal) that provides some of the Windows API on UNIX-like platforms, but is there something similar within the SAMATE tests?

**Comment by geoffw0** [comment]
> Is there a better way to tell what standard library the code is actually being compiled against?

SAMATE compiles against standard header files + a few of it's own header files, using the `_WIN32` macro to make a few choices about what to include here and there and make sure it compiles.  So if anything it looks like it's intended to be cross-platform.  But the code is plainly written with only Windows in mind in some places - for example, from one of the "good" cases:
```
wchar_t * data;

...

fwprintf(stdout, L"%s\n", data);
```

This is the kind of mess that motivates me not wanting to report these cases.

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**